pennylane/beta/tapes/tape.py

```diff
@@ -0,0 +1,776 @@
+                else:
+                    self._ops.append(obj)
+
+            elif isinstance(obj, MeasurementProcess):
+                # measurement process
+
+                if obj.return_type is qml.operation.Probability:
+                    self._obs.append((obj, obj))
+                    self._output_dim += 2 ** len(obj.wires)
+
+                elif "owns" in info:
+                    # TODO: remove the following line once devices
+                    # have been refactored to no longer use obs.return_type.
+                    # Monkeypatch the observable to have the same return
+                    # type as the measurement process.
+                    info["owns"].return_type = obj.return_type
+                    info["owns"].do_check_domain = False
+
+                    self._obs.append((obj, info["owns"]))
+                    self._output_dim += 1
+
+                    if obj.return_type is qml.operation.Sample:
+                        self.is_sampled = True
+
+            elif isinstance(obj, qml.operation.Observable) and "owner" not in info:
+                raise ValueError(f"Observable {obj} does not have a measurement type specified.")
+
+        self._update()
+
+    def _update_circuit_info(self):
+        """Update circuit metadata"""
+        self.wires = qml.wires.Wires.all_wires(
+            [op.wires for op in self.operations + self.observables]
+        )
+        self.num_wires = len(self.wires)
+
+    def _update_par_info(self):
+        """Update the parameter information dictionary"""
+        param_count = 0
+
+        for obj in self.operations + self.observables:
+
+            for p in range(len(obj.data)):
+                info = self._par_info.get(param_count, {})
+                info.update({"op": obj, "p_idx": p})
+
+                self._par_info[param_count] = info
+                param_count += 1
+
+    def _update_trainable_params(self):
+        """Set the trainable parameters"""
+        self._trainable_params = set(self._par_info)
+
+    def _update(self):
+        """Update all internal tape metadata regarding processed operations and observables"""
+        self._graph = None
+        self._update_circuit_info()
+        self._update_par_info()
+        self._update_trainable_params()
+
+    def _expand(self, stop_at=None):
+        """Expand all operations and tapes in the processed queue.
+
+        Args:
+            stop_at (Sequence[str]): Sequence of PennyLane operation or tape names.
+                An operation or tape appearing in this list of names is not expanded.
+        """
+        new_tape = self.__class__()
+        stop_at = stop_at or []
+
+        for op in self.operations:
+            if op.name in stop_at:
+                new_tape._ops.append(op)
+                continue
+
+            t = op if isinstance(op, QuantumTape) else op.expand()
+            new_tape._prep += t._prep
+            new_tape._ops += t._ops
+
+        new_tape._obs = self._obs
```

Comment:
This is a good point. Diagonalizing gates are subtly different from operation decompositions:

* With operations, we can expand them out and replace them directly with the decompositions, and no information is lost.

* With observables, we can't just wholesale replace the observable (e.g., `obs=PauliX -> ops+=Hadamard, obs=PauliZ`) because we are losing information. We also need to retain information regarding post-measurement classical processing (the eigenvalues of the original observable need to be preserved).

Let me think about the best way of incorporating this. Perhaps we simply add an additional tape list attribute, to keep track of any eigenvalue post-processing that needs to be done.

Reply:
I noticed that `QuantumTape` inherits from `AnnotatedQueue` :grinning:

Reply:
By the time expansion is called, however, the queue has been processed and the tape is no longer using/referring to the annotated queue!

Reply:
I've added support for this. It ended up being quite elegant in the end; simply adding a `MeasurementProcess.expand()` option, and including the measurement queue in the expansion logic.

Note that devices won't be able to make use of this until we refactor their logic; for now, they will continue to access `tape.diagonalizing_gates()`.

Reply:
:fire: